### PR TITLE
#58 Fix early VIATRA engine initialization in ViatraQueryAdapter

### PR DESCRIPTION
**Pull request by tamasborbas**
* provide new API to get engine optionally and be able to initialize it later
* make ViatraQueryAdapter#getEngine method deprecated

### CODE REVIEW

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 1]
```diff
@@ -23,29 +32,119 @@
  * 
  */
 public class ViatraQueryAdapter extends AdapterImpl {
+	public static final String MESSAGE_ENGINE_NOT_READY = 
+			"VIATRA engine is not available while the primary project is not loaded!";
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:36]:
```suggestion
			"Cannot initialize VIATRA Query Engine until the project is loaded.";
```

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/expressions/BinaryVQLExpression.java [hunk 2]
```diff
@@ -51,7 +53,10 @@ public BinaryVQLExpression(ParameterizedExpressionDescriptor descriptor, Express
 	}
 
 	public Object getValue(Element sourceParameter, ValueContext context) throws Exception {
-		ViatraQueryEngine engine = ViatraQueryAdapter.getAdapter(project).get().getEngine();
+		ViatraQueryEngine engine = ViatraQueryAdapter.getAdapter(project).orElseThrow(()-> {
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/expressions/BinaryVQLExpression.java:56]: I don't think it is a good idea to throw an exception in the getValue() method as it gets called by MagicDraw APIs and the exception trace would be highly visible in the logs. I think in such cases simply returning an empty match set (similar to when the Binary Query Result Provider couldn't be initialized) would be a better approach here.

@lunkpeter you have more experience with using these expressions, what do you think about its updates.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 3]
```diff
@@ -23,29 +32,119 @@
...
+	private ViatraQueryAdapter(Project project, Notifier[] notifiers) {
 		this.project = project;
+		this.notifiers = notifiers;
 	}
-	
+
+	/**
+	 * @return initialized engine
+	 * @throws ViatraQueryException if engine cannot be initialized yet
+	 * @deprecated Use {@link #getInitializedEngineChecked()} instead
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:55]:
```suggestion
	 * @deprecated Use {@link #getInitializedEngineChecked()} or {@link #getInitializedEngine()} instead
```

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 4]
```diff
@@ -23,29 +32,119 @@
...
+	 * Create a matcher for the VIATRA query engine of the adapter. If it cannot be created
+	 * immediately, the initialization of the matcher will be done when the engine is ready.
+	 * 
+	 * @param <M> The IQuerySpecification implementation specific type of the 
+	 * ViatraQueryMatcher which would be initialized by this method
+	 * @param querySpecification is the query for which an initialized matcher is needed
+	 * @return the matcher if it could haven been initialized or an empty optional if not
+	 */
+	public <M extends ViatraQueryMatcher<? extends IPatternMatch>> Optional<M> 
+			initializeQueryWhenEngineReady(IQuerySpecification<M> querySpecification) {
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:131]: I think the name of this method could simply be `requireMatcher` - together with the Optional return value it expresses that the matcher might not be initialized at the moment of the call.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 5]
```diff
@@ -23,29 +32,119 @@
...
+	private Optional<AdvancedViatraQueryEngine> initializeEngine() {
+		if(!isInitialized && ProjectUtilities.isLoaded(project.getPrimaryProject())) {
+			engine = Optional.of(createQueryEngine(project, notifiers));
+			notifiers = new Notifier[0];
+			isInitialized = true;
+		}
+		if(!isInitialized) {
+			LOGGER.warn(MESSAGE_ENGINE_NOT_READY);
+		} else {
+			initializationActions.forEach(action -> action.accept(engine.get()));
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:115]: We should wrap the initialization actions into a coalesceTraversal block otherwise it is possible that multiple model traversals might be triggered for the delayed initialization actions. See [VIATRA documentation](https://www.eclipse.org/viatra/documentation/query-api.html#_query_groups_and_coalescing_model_traversals) for details if necessary.

**Review by ujhelyiz — CHANGES_REQUESTED** [review]
In general the changes look right but I have a few comments we should address before merging this to master.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/expressions/BinaryVQLExpression.java (hunk 2 above)
Comment (lunkpeter) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/expressions/BinaryVQLExpression.java:56]: I agree completely. The expression executor does not handle internal exception particularly well, resulting in internal error notifications all across the UI

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 6]
```diff
@@ -23,29 +32,119 @@
...
+	}
+
+	/**
+	 * Try to initialize the VIATRA engine if it hasn!t been before and if the 
+	 * initialization is not possible, it throws a ViatraQueryException.
+	 * 
+	 * @return the initialized engine
+	 * @throws ViatraQueryException if initialized engine is not available yet
+	 */
+	public AdvancedViatraQueryEngine getInitializedEngineChecked() {
```
Comment (lunkpeter) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:69]: Why is this a separate public method? 
 - Does exactly the same as getEngine
 - Newly introduced so serves no purpose in backwards compatibility

What I'm guessing is to have an equivalent of the getEngine method following the new naming conventions, however imho it just makes the API confusing

If the purpose is to replace the current getEngine, we should specify ViatraQueryException in a proper throws block --> I know it is a runtime exception but the user needs to handle it from the calling code.

**Review by lunkpeter — CHANGES_REQUESTED** [review]

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 6 above)
Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:69]: It was me who suggested this to Tamás, and my reasoning was to have a consistent terminology of accessing the engines, and the old getEngine really did not fit. As you have noticed, the new method is aimed to be a clear replacement for the old one, but we did not want to break all existing code (hence we only deprecated the getEngine method).

About adding the throws declaration, I am not against it, but it really does not add a lot, and RuntimeExceptions are not expected to be added as a throws declaration.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/expressions/BinaryVQLExpression.java (hunk 2 above)
Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/expressions/BinaryVQLExpression.java]: Previously the optional was able to throw an exception (it wasn't checked before the get was called) and I thought that a more specific exception would be more useful but I understand your point so now I handle the uninitialized engine instead of throwing exceptions.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 4 above)
Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java]: I renamed the method according to your suggestion.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 5 above)
Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java]: I tried to use `coalesceTraversals` but I'm not sure if I used it in the correct way. Please, check the new implementation again.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 6 above)
Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: Would it be better if we handle the exception in the original `getEngine` and return `null` instead of throwing an exception?

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 7]
```diff
@@ -104,16 +107,31 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
...
+								return null;
+							});
+							initializationActions.clear();
+							notifiers = new Notifier[0];
+						} catch (InvocationTargetException ite) {
+							LOGGER.warn(MESSAGE_ENGINE_PREPARE_ACTION_ERROR);
+							engine = null;
+							thereWasException = true;
+						}
+						return !thereWasException;
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:129]: The control flow in this method is really strange (it might be correct, but it is strange). In case an exception was thrown during initialization, we state that the engine is not initialized throw away the engine created, but keep all the previously saved actions for a second try (do we hope that the exception will get resolved at that point?)...

Question: is this behavior consistent with what happens when the engine initialization throws a ViatraQueryException (of course a different error message is fine, but we should handle these errors the same way).

**Review by ujhelyiz — CHANGES_REQUESTED** [review]
While the coalesceTraversal is used as expected, the error handling does not seem correct to me.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 6 above)
Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: I don't think so. Returning null is a pattern we very much try to avoid here, especially in case of errors. Here the exception has a lot of added value: we explicitly say that something went wrong instead of the user figuring it out later when the engine got corrupted at a random model update.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 5 above)
Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java]: The `coalesceTraversals` is used correctly, but I have opened extra discussions with further issues.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 7 above)
Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:129]: I think it is possible that we cannot initialize the engine correctly at some point but we can later (based on what I have seen from MagicDraw).
And yes, it is consistent: we throw a ViatraQueryException if the VIATRA engine initialization is not successful which is true in two cases: a) if it cannot be created or b) the execution of an action fails.

**Review by ujhelyiz — APPROVED** [review]
Thanks, now it looks good to me.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 6 above)
Comment (lunkpeter) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: I think throwing a non-runtime exception would solve all of my concerns. I agree though that the circumstances are pretty much the definition for runtime exceptions, but I think the specific nature of the platform necessitates that the user of this API always handles the cases when the engine is not ready

Right now when migrating or using the non-optional method (which the user will do cause its simpler) the user is not enforced to handle situations where the engine is not ready. And in most cases in MD uncaught exceptions coming from certain integrated functionality can break the entire client. I'd rather have the users add some unnecesary try catch blocks.

Comment (lunkpeter) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: Also it is important to know that the exception is not thrown when the engine is corrupted due to irregularities, but rather when the user tries to use an engine before it is ready.

Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: The problem is that adding a non-runtime exception only to new methods would ensure users have to check for them. This would be an outlier for all existing APIs where it was a design decision to follow VIATRA's approach of always throwing runtime exceptions (and changing all the APIs would break compatibility too much).

On the other hand, I have no problem adding documentation to the Javadoc, stating that this method should only be used by users who can be sure that the project loading is finished. Otherwise, the user should rely on the method returning an optional. If you have a name idea that better suggests this constraint, I am OK with renaming it as well. Would such a compromise solve your concerns?

Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: What do you think about throwing a non-runtime exception? In our case, we know the problem, it can be handled by time and there is nothing fatal from which we cannot stand up. Maybe we can have a custom `UnsuccessfulViatraEngineInitializationException` or something like this.

Comment (lunkpeter) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: This is only a minor concern though I'm willing to let it go if it's documented properly. (which seems to be the case)

Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:72]: Based on the discussion above (and because I have found a typo) I suggest rewriting the Javadoc as follows (and I cannot really add it as an inline recommendation because it is split into multiple lines and I would like to add it in a reviewable text here (please check the Javadoc syntax, I was writing it as a text):

```
Returns an initialized VIATRA query engine stored in the adapter.
If a query engine was initialized earlier, that engine will be returned;
if no query engine was initialized earlier, a new one is created. If
the initialization fails, e.g. the engine was requested before the
model was loaded entirely, a {@link ViatraQueryException} is thrown.
</p>
This method should only be called if the engine initialization is
expected to be completed successfully, otherwise it is recommended
to use {@link #getInitializedEngine()},
{@link #getInitializedEngine(Consumer<AdvancedViatraQueryEngine>)}
or {@link #requireMatcher(IQuerySpecification<M>)} instead
as they are aimed to handle the initialization issues better.

@return the initialized engine
@throws ViatraQueryException if no query engine was initialized before and initialization failed
```

**Review by ujhelyiz — APPROVED** [review]
Looks good to me.

**Review by lunkpeter — APPROVED** [review]

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 8]
```diff
@@ -23,29 +33,154 @@
...
 	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
+	 * If not, the action will be registered in the list of actions which need to be done 
+	 * when the engine is ready and will be executed when the engine become available.
+	 * 
+	 * @param action the operation which would like to use the initialized VIATRA engine
+	 * @return the initialized VIATRA engine of it is available, otherwise empty Optional.
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine(Consumer<AdvancedViatraQueryEngine> action) {
```
Comment (debrecenics) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:112]: From user perspective, I assume we execute a non-blocking action even if the engine is not yet ready. This means the user does not really interested in whether the engine is ready or not. Hence, why would he need the return value of this method? I would suggest changing to `public void  executeActionWhenEngineInitialized(Consumer<AdvancedViatraQueryEngine> action)` without returning value.

Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:112]: Good catch. We had discussed this with @tamasborbas in person as well and he will provide a newer version that handles this case (and updates the API a bit).

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 9]
```diff
@@ -109,67 +107,79 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
 	 * @param action the operation which would like to use the initialized VIATRA engine
 	 * @return the initialized VIATRA engine of it is available, otherwise empty Optional.
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:108]: Remove the @returns statement

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 8 above)
Comment (debrecenics) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java]: I might also mark the method somehow with "async" (e.g. in the name) to make sure the users won't try to check the result of the action immediately. Or even adding a callback function to let the user know when the action is executed. But if it is against the naming convention or these imply unnecessary modification, I'm okay with the current change and with the entire PR.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 10]
```diff
@@ -23,29 +31,166 @@
...
+		return engine;
+	}
+	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
+	 * If not, the action will be registered in the list of actions which need to be done 
+	 * when the engine is ready and will be executed when the engine become available.
+	 * 
+	 * @param action the operation which would like to use the initialized VIATRA engine
+	 * @return the initialized VIATRA engine of it is available, otherwise empty Optional.
```
Comment (debrecenics) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:108]: Please remove this line from the JavaDoc.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 11]
```diff
@@ -23,29 +31,166 @@
...
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine() {
+		if(!isInitialized) {
+			initializeEngine();
+		}
+		return engine;
+	}
+	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:103]: I would reword this as follows:

Executes the give action on the engine associated with the current adapter. If the engine is not yet initialized, the action is saved and will be executed later when the engine gets initialized.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 12]
```diff
@@ -109,67 +107,79 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
...
+						e.getBaseIndex().coalesceTraversals(() -> {
+							initializationActions.forEach(action -> action.accept(e));
+							return null;
+						});
+						initializationActions.clear();
+						notifiers = new Notifier[0];
+					} catch (InvocationTargetException ite) {
+						// we can invalidate our engine because there is two option for this exception:
+						// 1. we throw something directly from the callable (but we don't do this)
+						// 2. the engine goes to tainted and throws something what is wrapped into this -> from this point the engine is unusable
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:136]:
```suggestion
						// 2. the engine got tainted because of an internal error, making the engine unusable
```

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 13]
```diff
@@ -109,67 +107,79 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
...
+					try {
+						e.getBaseIndex().coalesceTraversals(() -> {
+							initializationActions.forEach(action -> action.accept(e));
+							return null;
+						});
+						initializationActions.clear();
+						notifiers = new Notifier[0];
+					} catch (InvocationTargetException ite) {
+						// we can invalidate our engine because there is two option for this exception:
+						// 1. we throw something directly from the callable (but we don't do this)
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:135]:
```suggestion
						// 1. an exception is thrown directly by the action (which is not supported)
```

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 14]
```diff
@@ -109,67 +107,79 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
 	 * @param action the operation which would like to use the initialized VIATRA engine
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:107]:
```suggestion
	 * @param action the operation which would like to use the initialized VIATRA engine - the action should not throw an exception during execution
```

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 15]
```diff
@@ -109,67 +107,79 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
...
+							initializationActions.forEach(action -> action.accept(e));
+							return null;
+						});
+						initializationActions.clear();
+						notifiers = new Notifier[0];
+					} catch (InvocationTargetException ite) {
+						// we can invalidate our engine because there is two option for this exception:
+						// 1. we throw something directly from the callable (but we don't do this)
+						// 2. the engine goes to tainted and throws something what is wrapped into this -> from this point the engine is unusable
+						// so even if we execute some action later than the  real initialization of the engine,
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:137]:
```suggestion
```

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java [hunk 16]
```diff
@@ -109,67 +107,79 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
...
+						// the engine will brake down and can't be used anymore
```
Comment (ujhelyiz) [inline com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java:138]:
```suggestion
```

**Review by ujhelyiz — CHANGES_REQUESTED** [review]

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 8 above)
Comment (ujhelyiz) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java]: This method is strange, because it might be either sync (if the engine is already initialized) or async (if the engine is not initialized yet), so the _async postfix would be misleading.

Your original recommendation of `whenEngineInitialized` is really too long; the name `executeActionOnEngine` is the best way, as the action cannot be executed when the engine is not initialized yet; but there is only a single engine associated with each adapter, so this should be clear what is it referring to.

File: com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java (hunk 11 above)
Comment (tamasborbas) [reply com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java]: I've replaced the previous JavaDoc with your suggestion

**Review by ujhelyiz — APPROVED** [review]
Looks reasonable to me.

**Review by debrecenics — APPROVED** [review]
